hash_map.h

```diff
@@ -0,0 +1,170 @@
+    }
+
+    const_iterator begin() const noexcept {
+        return nodes.begin();
+    }
+
+    iterator end() noexcept {
+        return nodes.end();
+    }
+
+    const_iterator end() const noexcept {
+        return nodes.end();
+    }
+
+    HashMap(Hash hasher = Hash()) : size_table(size_default), hasher(hasher), table(size_default) {}
+
+    HashMap(std::initializer_list<value_type_nonconst> init, Hash hasher = Hash()) : HashMap() {
+        fill(init);
+    }
+
+    template<class IteratorType>
+    HashMap(IteratorType bgn, IteratorType end, Hash hasher = Hash()) : HashMap() {
+        for (auto i = bgn; i != end; i++) {
+            insert(*i);
+        }
+    }
+
+    std::size_t size() const {
+        return nodes.size();
+    }
+
+    bool empty() const {
+        return nodes.size() == 0;
+    }
+
+    const Hash hash_function() const {
+        return hasher;
+    }
+
+    void insert(value_type insertable) {
+        KeyType user_key = insertable.first;
+        if (find(user_key) != nodes.end()) {
+            return;
+        }
+        std::size_t cur_hash = hasher(user_key) % size_table;
+        nodes.push_back(insertable);
+        iterator cur_it = std::prev(nodes.end(), 1);
+        table[cur_hash].push_back(cur_it);
+        if (table[cur_hash].size() > size_table / size_default) {
+            rebuilt();
+        }
+    }
+
+    void erase(KeyType user_key) {
+        std::size_t cur_hash = hasher(user_key) % size_table;
+        if (find(user_key) != nodes.end()) {
+            auto cur_it = table[cur_hash].begin();
+            for (auto i : table[cur_hash]) {
+                if (i->first == user_key) {
+                    nodes.erase(i);
+                    table[cur_hash].erase(cur_it);
+                    break;
+                }
+                cur_it++;
+            }
+        }
+    }
+
+    iterator find(KeyType user_key) noexcept {
+        std::size_t cur_hash = hasher(user_key) % size_table;
+        for (auto iter : table[cur_hash]) {
+            KeyType cur_key = iter->first;
+            if (cur_key == user_key) {
+                return iter;
+            }
+        }
+        return nodes.end();
+    }
+
+    const_iterator find(KeyType user_key) const noexcept {
+        std::size_t cur_hash = hasher(user_key) % size_table;
+        for (auto iter : table[cur_hash]) {
+            KeyType cur_key = iter->first;
+            if (cur_key == user_key) {
+                return iter;
+            }
+        }
+        return nodes.end();
+    }
+
+    void clear() {
+        table.clear();
+        size_table = size_default;
+        table.resize(size_default);
+        nodes.clear();
+    }
+
+    const ValueType &at(KeyType user_key) const {
+        if (find(user_key) == nodes.end()) {
+            throw std::out_of_range("");
+        }
+        return find(user_key)->second;
+    }
+
+    ValueType &operator[](const KeyType user_key) {
+        if (find(user_key) == nodes.end()) {
```

Comment:
You will call find three times on every call of operator[].

Reply:
Same comment as in the previous review.

Reply:
fixed